internal/source/git.go

```diff
@@ -2,145 +2,91 @@ package source
-			util.CoreOwnerKindKey: bundle.Kind,
-			util.CoreOwnerNameKey: bundle.Name,
-		})
-		pod.SetOwnerReferences([]metav1.OwnerReference{*controllerRef})
-		pod.Spec.AutomountServiceAccountToken = &automountServiceAccountToken
-		pod.Spec.RestartPolicy = corev1.RestartPolicyNever
-
-		if len(pod.Spec.InitContainers) != 1 {
-			pod.Spec.InitContainers = make([]corev1.Container, 1)
-		}
+	if gitsource.Ref.Branch != "" {
+		cloneOpts.ReferenceName = plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", gitsource.Ref.Branch))
+		cloneOpts.SingleBranch = true
+		cloneOpts.Depth = 1
+	} else if gitsource.Ref.Tag != "" {
+		cloneOpts.ReferenceName = plumbing.ReferenceName(fmt.Sprintf("refs/tags/%s", gitsource.Ref.Tag))
+		cloneOpts.SingleBranch = true
+		cloneOpts.Depth = 1
+	}
 
-		// Note: initContainer so we can ensure the repository has been cloned
-		// at the bundle.Spec.Source.Git.Ref before we unpack the Bundle contents
-		// that are stored in the repository.
-		pod.Spec.InitContainers[0].Name = "clone-repository"
-		// r.GitClientImage configures which git-based container image to use to clone the provided repository
-		// r.GitClientImage currently defaults to alpine/git:v2.32.0
-		pod.Spec.InitContainers[0].Image = g.GitClientImage
-		pod.Spec.InitContainers[0].ImagePullPolicy = corev1.PullIfNotPresent
-		cmd, err := git.CloneCommandFor(*bundle.Spec.Source.Git)
-		if err != nil {
-			return err
-		}
-		pod.Spec.InitContainers[0].Command = []string{"/bin/sh", "-c", cmd}
-		pod.Spec.InitContainers[0].VolumeMounts = []corev1.VolumeMount{{Name: "bundle", MountPath: "/bundle"}}
+	// Clone
+	repo, err := git.CloneContext(ctx, memory.NewStorage(), memfs.New(), &cloneOpts)
+	if err != nil {
+		return nil, fmt.Errorf("bundle unpack git clone error: %w", err)
+	}
+	wt, err := repo.Worktree()
+	if err != nil {
+		return nil, fmt.Errorf("bundle unpack error: %w", err)
+	}
 
-		if len(pod.Spec.Containers) != 1 {
-			pod.Spec.Containers = make([]corev1.Container, 1)
+	// Checkout commit
+	if gitsource.Ref.Commit != "" {
+		commitHash := plumbing.NewHash(gitsource.Ref.Commit)
+		if err := wt.Reset(&git.ResetOptions{
+			Commit: commitHash,
+			Mode:   git.HardReset,
+		}); err != nil {
+			return nil, fmt.Errorf("checkout commit %q: %w", commitHash.String(), err)
 		}
+	}
 
-		pod.Spec.Containers[0].Name = gitBundleUnpackContainerName
-		pod.Spec.Containers[0].Image = g.UnpackImage
-		pod.Spec.Containers[0].ImagePullPolicy = corev1.PullIfNotPresent
-		pod.Spec.Containers[0].Command = []string{"/unpack", "--bundle-dir", "/bundle"}
-		pod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{Name: "bundle", MountPath: "/bundle"}}
+	var bundleFS fs.FS = &billyFS{wt.Filesystem}
 
-		pod.Spec.Volumes = []corev1.Volume{
-			{Name: "bundle", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
+	// Subdirectory
+	if gitsource.Directory != "" {
+		directory := filepath.Clean(gitsource.Directory)
+		if directory[:3] == "../" {
```

Comment:
This is a good catch and maybe something maybe we could move to the schema for validation in a follow-up? Based on the `Clean` documentation though, I think it replaces all relative paths so this condition would never be hit? https://pkg.go.dev/path/filepath#Clean

Also, would we want to check for `/` as an invalid directory as well?

Reply:
I'll add the check for '/'.  I'll create a follow up for the webhook changes.  Thanks!

Reply:
My understanding is that `Clean` can't do anything about a path with `../` prefixes. That makes sense because `Clean` doesn't assume CWD when it's working.

In cases where `../` is in the middle of a path, it _can_ clean those up because it can just remove the leading directories from the path.